hail/src/main/scala/is/hail/methods/MatrixExportColsParallel.scala

```diff
@@ -0,0 +1,143 @@
+
+import is.hail.annotations.{UnsafeIndexedSeq, UnsafeRow}
+import is.hail.expr.TableAnnotationImpex
+import is.hail.expr.ir.MatrixValue
+import is.hail.expr.ir.functions.MatrixToValueFunction
+import is.hail.expr.types.MatrixType
+import is.hail.expr.types.virtual.{TVoid, Type}
+import is.hail.utils._
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.Row
+
+case class MatrixExportColsParallel(parallelism: Int, path: String, bgzip: Boolean) extends MatrixToValueFunction {
+  def typ(childType: MatrixType): Type = TVoid
+
+  def execute(mv: MatrixValue): Any = {
+
+    val hConf = mv.sparkContext.hadoopConfiguration
+
+    hConf.delete(path, recursive = true) // overwrite by default
+
+    val allColValuesJSON = mv.colValues.value.map(TableAnnotationImpex.exportAnnotation(_, mv.typ.colType)).toArray
+
+    info(s"exporting ${ mv.nCols } files in batches of $parallelism...")
+    val nBatches = (mv.nCols + parallelism - 1) / parallelism
+    val resultFiles = (0 until nBatches).flatMap { batch =>
+      val startIdx = parallelism * batch
+      val nCols = mv.nCols
+      val endIdx = math.min(nCols, parallelism * (batch + 1))
+
+      info(s"on batch ${ batch + 1 } of ${ nBatches }, columns $startIdx to ${ endIdx - 1 }...")
+
+      val d = digitsNeeded(mv.rvd.getNumPartitions)
+
+      val rvType = mv.rvd.rowPType
+      val entriesIdx = MatrixType.getEntriesIndex(rvType)
+      val entryArrayType = MatrixType.getEntryArrayType(rvType)
+      val entryType = MatrixType.getEntryType(rvType)
+
+      val partFileBase = path + "/tmp/"
+
+      val hConfBc = mv.sparkContext.broadcast(new SerializableHadoopConfiguration(hConf))
+
+      val extension = if (bgzip) ".tsv.bgz" else ".tsv"
+
+      val colValuesJSON = mv.sparkContext.broadcast(
+        (startIdx until endIdx)
+          .map(allColValuesJSON)
+          .toArray)
+
+      val partFolders = mv.rvd.crdd.mapPartitionsWithIndex { (i, it) =>
+
+        val partFolder = partFileBase + partFile(d, i, TaskContext.get())
+
+        val fileHandles = Array.tabulate(endIdx - startIdx) { j =>
+          new OutputStreamWriter(hConfBc.value.value.unsafeWriter(partFolder + "/" + j.toString + extension), "UTF-8")
+        }
+
+        if (i == 0) {
+          // write headers
+          val header = (
+            rvType.fieldNames.filter(_ != MatrixType.entriesIdentifier) ++ entryType.fieldNames
+            ).mkString("\t")
+
+          fileHandles.zipWithIndex.foreach { case (f, jj) =>
+            f.write('#')
+            f.write(colValuesJSON.value(jj))
+            f.write('\n')
+            f.write(header)
+            f.write('\n')
+          }
+        }
+
+        it.foreach { rv =>
+
+          val entriesArray = new UnsafeIndexedSeq(entryArrayType, rv.region, rvType.loadField(rv, entriesIdx))
+
+          val fullRow = new UnsafeRow(rvType, rv)
+
+          val rowFieldStrs = (0 until rvType.size)
+            .filter(_ != entriesIdx)
+            .map { rowFieldIdx =>
```

Comment:
Do we want the full row or just the row_key?

Reply:
Full thing. This is totally general.